snuba/query/processors/uuid_column_processor.py

```diff
@@ -0,0 +1,150 @@
+from typing import List
+import uuid
+
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.query.conditions import (
+    binary_condition,
+    ConditionFunctions,
+    FUNCTION_TO_OPERATOR,
+    is_condition,
+)
+from snuba.query.logical import Query
+from snuba.query.matchers import (
+    AnyOptionalString,
+    Or,
+    Param,
+    String,
+)
+from snuba.query.matchers import Column as ColumnMatch
+from snuba.query.matchers import FunctionCall as FunctionCallMatch
+from snuba.query.matchers import Literal as LiteralMatch
+from snuba.query.processors import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+
+
+class UUIDColumnProcessor(QueryProcessor):
+    """
+    If a condition is being performed on a column that stores UUIDs (as defined in the constructor)
+    then change the condition to use a proper UUID instead of a string.
+    """
+
+    def formatted_uuid_pattern(self, suffix: str = "") -> FunctionCallMatch:
+        return FunctionCallMatch(
+            Param("format_alias" + suffix, AnyOptionalString()),
+            String("replaceAll"),
+            (
+                FunctionCallMatch(
+                    None,
+                    String("toString"),
+                    (
+                        Param(
+                            "formatted_uuid_column" + suffix,
+                            ColumnMatch(None, None, self.__uuid_column_match),
+                        ),
+                    ),
+                ),
+            ),
+            with_optionals=True,
+        )
+
+    def uuid_column_pattern(self, suffix: str = "") -> Param:
+        return Param(
+            "uuid_column" + suffix, ColumnMatch(None, None, self.__uuid_column_match)
+        )
+
+    def __init__(self, uuid_columns: List[str]) -> None:
+        self.__unique_uuid_columns = set(uuid_columns)
+        self.__uuid_column_match = Or(
+            [String(u_col) for u_col in self.__unique_uuid_columns]
+        )
+        self.uuid_in_condition = FunctionCallMatch(
+            None,
+            Or((String(ConditionFunctions.IN), String(ConditionFunctions.NOT_IN))),
+            (
+                Or((self.uuid_column_pattern(), self.formatted_uuid_pattern())),
+                Param("params", FunctionCallMatch(None, String("tuple"), None)),
+            ),
+        )
+        self.uuid_condition = FunctionCallMatch(
+            None,
+            Or([String(op) for op in FUNCTION_TO_OPERATOR]),
+            (
+                Or(
+                    (
+                        Param("literal_0", LiteralMatch(None, AnyOptionalString())),
+                        self.uuid_column_pattern("_0"),
+                        self.formatted_uuid_pattern("_0"),
+                    )
+                ),
+                Or(
+                    (
+                        Param("literal_1", LiteralMatch(None, AnyOptionalString())),
+                        self.uuid_column_pattern("_1"),
```

Comment:
I am not sure why you need to match expressions like:
```
Column(None, None, "something")
=
Literal(...)
```

Those should not be optimized, can't you just leave them as it is and not add them to the matcher so they would be skipped ?

Reply:
There's nothing about context promotion here. This processor does two things: If a condition is using a uuid column, as specified in the constructor, convert the literal to be in UUID format to correctly match. If the uuid column has been wrapped with `replaceAll(...`, unwrap it. This processor is useless without changing the literal, since those are not passed as valid UUIDs from Sentry. Something has to convert those values to UUIDs.

Reply:
Oh I see it now, the column you need to transform here is the one that was inside the wrapped expression. This makes sense.